File: sample/src/main/java/com/polidea/rxandroidble2/sample/util/LocationPermission.java
package com.polidea.rxandroidble2.sample.util;

import android.app.Activity;
import android.content.pm.PackageManager;
import androidx.core.app.ActivityCompat;
import com.polidea.rxandroidble2.RxBleClient;

public class LocationPermission {

    private LocationPermission() {
        // Utility class
    }

    private static final int REQUEST_PERMISSION_BLE_SCAN = 9358;

    public static void requestLocationPermission(final Activity activity, final RxBleClient client) {
        ActivityCompat.requestPermissions(
                activity,
                /*
                 * the below would cause a ArrayIndexOutOfBoundsException on API < 23. Yet it should not be called then as runtime
                 * permissions are not needed and RxBleClient.isScanRuntimePermissionGranted() returns `true`
                 */
                new String[]{client.getRecommendedScanRuntimePermissions()[0]},
                REQUEST_PERMISSION_BLE_SCAN
        );
    }

    public static boolean isRequestLocationPermissionGranted(final int requestCode, final String[] permissions,
                                                             final int[] grantResults, RxBleClient client) {
        if (requestCode != REQUEST_PERMISSION_BLE_SCAN) {
            return false;
        }

        String[] recommendedScanRuntimePermissions = client.getRecommendedScanRuntimePermissions();
        for (int i = 0; i < permissions.length; i++) {
            for (String recommendedScanRuntimePermission : recommendedScanRuntimePermissions) {
                if (permissions[i].equals(recommendedScanRuntimePermission)
                        && grantResults[i] == PackageManager.PERMISSION_GRANTED) {
                    return true;
                }
            }
        }

        return false;
    }
}

File: sample/src/main/java/com/polidea/rxandroidble2/sample/util/LocationPermission.kt
package com.polidea.rxandroidble2.sample.util

import android.Manifest
import android.Manifest.permission
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat

private const val REQUEST_PERMISSION_COARSE_LOCATION = 9358

internal fun Context.checkLocationPermissionGranted(): Boolean =
    ContextCompat.checkSelfPermission(this, permission.ACCESS_COARSE_LOCATION) ==
            PackageManager.PERMISSION_GRANTED

internal fun Activity.requestLocationPermission() {
    ActivityCompat.requestPermissions(
        this,
        arrayOf(permission.ACCESS_COARSE_LOCATION),
        REQUEST_PERMISSION_COARSE_LOCATION
    )
}

internal fun isRequestLocationPermissionGranted(
    requestCode: Int,
    permissions: Array<String>,
    grantResults: IntArray
): Boolean {
    if (requestCode == REQUEST_PERMISSION_COARSE_LOCATION) {
        permissions.forEachIndexed { index, permission ->
            if (
                permission == Manifest.permission.ACCESS_COARSE_LOCATION &&
                grantResults[index] == PackageManager.PERMISSION_GRANTED
            ) {
                return true
            }
        }
    }
    return false
}
